graph.go

```diff
@@ -11,197 +11,162 @@ type Node interface {
-	// determining things like if a graph is fully connected. The caller is
-	// free to modify this list. Implementations should construct a new list
-	// and not return internal representation.
-	NodeList() []Node
-
-	// Neighbors returns all nodes connected by any edge to this node.
-	Neighbors(Node) []Node
+// Out specifies a graph that can provide a list of outgoing edges
+// for an input node. The outgoing edges are expected to have Head().ID()
+// == n.ID().
+type Out interface {
+	Out(n Node) []DirectedEdge
+}
 
-	// EdgeBetween returns an edge between node and neighbor such that
-	// Head is one argument and Tail is the other. If no
-	// such edge exists, this function returns nil.
-	EdgeBetween(node, neighbor Node) Edge
+// In specifies a graph that can provide a list of incoming edges
+// for an input node. The incoming edges are expected to have
+// Tail().ID() == n.ID()
+type In interface {
+	In(n Node) []DirectedEdge
 }
 
-// Directed graphs are characterized by having seperable Heads and Tails in their edges.
-// That is, if node1 goes to node2, that does not necessarily imply that node2 goes to node1.
+// InOut means that a graph can generate predecessor and
+// successors, the InOut function has no guarantees about which is head
+// or tail.
 //
-// While it's possible for a directed graph to have fully reciprocal edges (i.e. the graph is
-// symmetric) -- it is not required to be. The graph is also required to implement Graph
-// because in many cases it can be useful to know all neighbors regardless of direction.
-type DirectedGraph interface {
-	Graph
-	// Successors gives the nodes connected by OUTBOUND edges.
-	// If the graph is an undirected graph, this set is equal to Predecessors.
-	Successors(Node) []Node
+// Note that this interface technically embeds EdgeProvider, but doesn't
+// explicitly do so for semantic reasons.
+type InOut interface {
+	In
+	Out
+	InOut(n Node) []DirectedEdge
+	Degree(n Node) uint
+}
 
-	// EdgeTo returns an edge between node and successor such that
-	// Head returns node and Tail returns successor, if no
-	// such edge exists, this function returns nil.
-	EdgeTo(node, successor Node) Edge
+// Provides the adjacent nodes to a given node.
+// This is for convenience, and could be derived from either
+// InOut or EdgeProvider
+type NeighborProvider interface {
+	Neighbors(n Node) []Node
+	Degree(n Node) uint
+}
 
-	// Predecessors gives the nodes connected by INBOUND edges.
-	// If the graph is an undirected graph, this set is equal to Successors.
-	Predecessors(Node) []Node
+// EdgeProvider is a generic method that provides undirected edges coming
+// to/from a node. For directed cases, use the InOut interface.
+type EdgeProvider interface {
+	Edges(n Node) []UndirectedEdge
+	Degree(n Node) uint
 }
 
-// Returns all undirected edges in the graph
-type EdgeLister interface {
-	EdgeList() []Edge
+type FiniteGraph interface {
```

Comment:
This is at the centre of something that has not been properly discussed; the goals and non-goals of the graph package.

All our graphs are finite. We have a contract that our nodes are uniquely defined in the context of the graph they are being used in. This limits the number of nodes we can have to max(uint) even if we are not returning a slice of the graph's nodes.

I think that by trying to accommodate true infinite graphs we are significantly weakening the value of the package for more common uses. There are far better languages for dealing with theoretical graphs. Handling infinite graphs (not to say implicitly defined or dynamically generated graphs) should be a non-goal for the package.